x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/StringMatcher.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    public static StringMatcher of(Iterable<String> patterns) {
+        return StringMatcher.builder().includeAll(patterns).build();
+    }
+
+    public static StringMatcher of(String... patterns) {
+        return StringMatcher.builder().includeAll(patterns).build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public String toString() {
+        return description;
+    }
+
+    @Override
+    public boolean test(String s) {
+        return predicate.test(s);
+    }
+
+    @Override
+    public StringMatcher or(Predicate<? super String> other) {
+        Objects.requireNonNull(other);
+        return new StringMatcher(description + "|" + other, this.predicate.or(other));
+    }
+
+    @Override
+    public StringMatcher and(Predicate<? super String> other) {
+        return this.and(String.valueOf(other), other);
+    }
+
+    public StringMatcher and(String otherDescription, Predicate<? super String> otherPredicate) {
+        Objects.requireNonNull(otherPredicate);
+        return new StringMatcher(this.description + "&" + otherDescription, this.predicate.and(otherPredicate));
+    }
+
+    public static class Builder {
+        private final List<String> allText = new ArrayList<>();
+        private final Set<String> exactMatch = new HashSet<>();
+        private final Set<String> nonExactMatch = new LinkedHashSet<>();
+
+        public Builder include(String pattern) {
+            allText.add(pattern);
+            if (pattern.startsWith("/") || pattern.contains("*") || pattern.contains("?")) {
+                nonExactMatch.add(pattern);
+            } else {
+                exactMatch.add(pattern);
+            }
+            return this;
+        }
+
+        public Builder includeAll(String... patterns) {
+            for (String pattern : patterns) {
+                include(pattern);
+            }
+            return this;
+        }
+
+        public Builder includeAll(Iterable<String> patterns) {
+            for (String pattern : patterns) {
+                include(pattern);
+            }
+            return this;
+        }
+
+        public StringMatcher build() {
+            if (allText.isEmpty()) {
+                return MATCH_NOTHING;
+            }
+
+            final String description = describe(allText);
+            if (exactMatch.isEmpty()) {
+                return new StringMatcher(description, buildAutomataPredicate(nonExactMatch));
+            }
+            if (nonExactMatch.isEmpty()) {
+                return new StringMatcher(description, buildExactMatchPredicate(exactMatch));
+            }
+            final Predicate<String> predicate = buildExactMatchPredicate(exactMatch).or(buildAutomataPredicate(nonExactMatch));
+            return new StringMatcher(description, predicate);
+        }
+
+        private String describe(List<String> strings) {
+            if (strings.size() == 1) {
+                return strings.get(0);
```

Comment:
Yes, it is. The 250 character limit is intentionally soft. It's there to prevent us generating 10kb of `toString` in logs/error messages, but I don't think we actually care about 250 vs 255 or even 300 characters.